From a51ae713b266b63c44c046031d51e87706b5c1f2 Mon Sep 17 00:00:00 2001
From: Hongzhang Yang <Hongzhang.Yang@freescale.com>
Date: Wed, 10 Jul 2013 17:48:03 +0800
Subject: [PATCH 237/280] ENGR00272337 Improve MJPEG line buffer mode

VPU lib v5.4.16

Align line buffer mode with ring buffer mode in JpegDecodeHeader:
- Do not return 0 if unexpected marker is found, just skip it
- Return -3 if headers are not complete

Signed-off-by: Hongzhang Yang <Hongzhang.Yang@freescale.com>
---
 vpu/vpu_lib.c  | 4 ++--
 vpu/vpu_lib.h  | 2 +-
 vpu/vpu_util.c | 7 +++----
 3 files changed, 6 insertions(+), 7 deletions(-)

diff --git a/vpu/vpu_lib.c b/vpu/vpu_lib.c
index fd45a86..4f89ef3 100644
--- a/vpu/vpu_lib.c
+++ b/vpu/vpu_lib.c
@@ -212,7 +212,7 @@ int vpu_WaitForInt(int timeout_in_ms)
 			IOClkGateSet(true);
 			if (VpuReadReg(BIT_BUSY_FLAG)) {
 				if (ret == 0)
-					warn_msg("intr received but VPU is still busy\n");
+					dprintf(4, "intr received but VPU is still busy\n");
 				ret = -1;
 			}
 			else
@@ -3729,7 +3729,7 @@ RetCode vpu_DecStartOneFrame(DecHandle handle, DecParam * param)
 			if (val == 0) {
 				UnlockVpu(vpu_semap);
 				return RETCODE_FAILURE;
-			} else if (val == -1) {
+			} else if ((val == -1) || (val == -3)) {
 				UnlockVpu(vpu_semap);
 				return RETCODE_JPEG_BIT_EMPTY;
 			}
diff --git a/vpu/vpu_lib.h b/vpu/vpu_lib.h
index 09f893a..9f57aea 100644
--- a/vpu/vpu_lib.h
+++ b/vpu/vpu_lib.h
@@ -663,7 +663,7 @@ typedef struct vpu_versioninfo {
  * v4.2.2 [2008.09.03] support encoder on MX51
  * v4.0.2 [2008.08.21] add the IOClkGateSet() for power saving.
  */
-#define VPU_LIB_VERSION_CODE	VPU_LIB_VERSION(5, 4, 15)
+#define VPU_LIB_VERSION_CODE	VPU_LIB_VERSION(5, 4, 16)
 
 extern unsigned int system_rev;
 
diff --git a/vpu/vpu_util.c b/vpu/vpu_util.c
index 5171b61..56fc7c9 100644
--- a/vpu/vpu_util.c
+++ b/vpu/vpu_util.c
@@ -2849,7 +2849,7 @@ int JpegDecodeHeader(DecInfo *pDecInfo)
 	for (;;) {
 		if (find_start_code(jpg) == 0) {
 			ret = -1;
-			err_msg("err in find start code\n");
+			dprintf(4, "err in find start code\n");
 			goto DONE_DEC_HEADER;
 		}
 
@@ -2926,6 +2926,7 @@ int JpegDecodeHeader(DecInfo *pDecInfo)
 			break;
 		case EOI_Marker:
 			ret = -3;
+			dprintf(4, "met EOI\n");
 			goto DONE_DEC_HEADER;
 		default:
 			switch (code & 0xFFF0) {
@@ -2945,8 +2946,6 @@ int JpegDecodeHeader(DecInfo *pDecInfo)
 				}
 			default:
 				dprintf(4, "code = [%x]\n", code);
-				if (jpg->lineBufferMode)
-					return 0;
 			}
 			break;
 		}
@@ -2969,7 +2968,7 @@ DONE_DEC_HEADER:
 
 	if (!jpg->ecsPtr || ret == -3) {
 		if (pDecInfo->jpgInfo.lineBufferMode)
-			return 0;
+			return -3;
 		else {
 			/* Skip the bitstream to EOI if EOI marker is found */
 			jpg->frameOffset += get_bits_count(&jpg->gbc) / 8 + 2;
-- 
1.8.0

